src/passes/LocalSubtyping.cpp

```diff
@@ -35,29 +36,60 @@ struct LocalSubtyping : public WalkerPass<PostWalker<LocalSubtyping>> {
 
   Pass* create() override { return new LocalSubtyping(); }
 
-  // Shared code to find all sets or gets for each local index. Returns a vector
-  // that maps local indexes => vector of LocalGet*|LocalSet* expressions.
-  template<typename T>
-  std::vector<std::vector<T*>> getLocalOperations(Function* func) {
-    std::vector<std::vector<T*>> ret;
-    ret.resize(func->getNumLocals());
-    FindAll<T> operations(func->body);
-    for (auto* operation : operations.list) {
-      ret[operation->index].push_back(operation);
-    }
-    return ret;
-  }
-
   void doWalkFunction(Function* func) {
     if (!getModule()->features.hasGC()) {
       return;
     }
 
-    auto varBase = func->getVarIndexBase();
     auto numLocals = func->getNumLocals();
 
-    auto setsForLocal = getLocalOperations<LocalSet>(func);
-    auto getsForLocal = getLocalOperations<LocalGet>(func);
+    // Compute the local graph. We need to get the list of gets and sets for
+    // each local, so that we can do the analysis. For non-nullable locals, we
+    // also need to know when the default value of a local is used: if so then
+    // we cannot change that type, as if we change the local type to
+    // non-nullable then we'd be accessing the default, which is not allowed.
+    //
+    // TODO: Optimize this, as LocalGraph computes more than we need, and on
+    //       more locals than we need.
+    LocalGraph localGraph(func);
+
+    // For each local index, compute all the the sets and gets.
+    std::vector<std::vector<LocalSet*>> setsForLocal(numLocals);
+    std::vector<std::vector<LocalGet*>> getsForLocal(numLocals);
+
+    for (auto& kv : localGraph.locations) {
+      auto* curr = kv.first;
+      if (auto* set = curr->dynCast<LocalSet>()) {
+        setsForLocal[set->index].push_back(set);
+      } else {
+        auto* get = curr->cast<LocalGet>();
+        getsForLocal[get->index].push_back(get);
+      }
+    }
+
+    // Find which vars use the default value, if we allow non-nullable locals.
+    //
+    // If that feature is not enabled, then we can safely assume that the
+    // default is never used - the default would be a null value, and the type
+    // of the null does not really matter as all nulls compare equally, so we do
+    // not need to worry.
```

Comment:
It's not like we're replacing it with some random other null, either, right? The new default null will always be a subtype of the original default null. This would even work in a hypothetical future where you could specify some other non-null default value for locals.